metagraph/src/graph/graph_extensions/node_rc.cpp

```diff
@@ -0,0 +1,274 @@
+    ips4o::parallel::sort(mapping.begin(), mapping.end(), utils::LessFirst(),
+                          get_num_threads());
+
+    std::vector<uint64_t> map;
+    map.reserve(mapping.size() * 2);
+
+    rc_ = Indicator([&](const auto &callback) {
+                        for (const auto &[node, rc] : mapping) {
+                            callback(node);
+                            map.emplace_back(rc.first);
+                            map.emplace_back(rc.second);
+                        }
+                    },
+                    graph.max_index() + 1,
+                    mapping.size());
+
+    mapping = decltype(mapping)();
+
+    logger->trace("Compressing reverse complement map");
+    mapping_ = { std::move(map) };
+}
+
+void NodeRC::call_outgoing_from_rc(node_index node,
+                                   const std::function<void(node_index)> &callback) const {
+    //        lshift    rc
+    // AGCCAT -> *AGCCA -> TGGCT*
+    assert(graph_);
+
+    if (const auto *dbg_succ_ = dynamic_cast<const DBGSuccinct*>(graph_)) {
+        //   AGAGGATCTCGTATGCCGTCTTCTGCTTGAG
+        //-> AGAGGATCTCGTATGCCGTCTTCTGCTTGA
+        //-> TCAAGCAGAAGACGGCATACGAGATCCTCT
+        const BOSS &boss = dbg_succ_->get_boss();
+        edge_index rc_edge = 0;
+
+        if (rc_.size()) {
+            if (auto rank = rc_.conditional_rank1(node))
+                rc_edge = mapping_[(rank - 1) * 2];
+        } else {
+            std::string rev_seq = dbg_succ_->get_node_sequence(node).substr(0, boss.get_k());
+            if (rev_seq[0] == BOSS::kSentinel)
+                return;
+
+            ::reverse_complement(rev_seq.begin(), rev_seq.end());
+            auto encoded = boss.encode(rev_seq);
+            auto [edge, edge_2, end] = boss.index_range(encoded.begin(), encoded.end());
+
+            if (end == encoded.end()) {
+                assert(edge == edge_2);
+                rc_edge = edge;
+            }
+        }
+
+        if (!rc_edge)
+            return;
+
+        boss.call_outgoing(rc_edge, [&](edge_index adjacent_edge) {
+            assert(dbg_succ_);
+            node_index prev = dbg_succ_->boss_to_kmer_index(adjacent_edge);
+            if (prev != DeBruijnGraph::npos)
+                callback(prev);
+        });
+
+        return;
+    }
+
+    // Do the checks by directly mapping the sequences of the desired k-mers.
+    // For non-DBGSuccinct graphs, this should be fast enough.
+    std::string rev_seq = graph_->get_node_sequence(node).substr(0, graph_->get_k() - 1);
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    rev_seq.push_back('\0');
+
+    const auto &alphabet = graph_->alphabet();
+    for (size_t c = 0; c < alphabet.size(); ++c) {
+        if (alphabet[c] != boss::BOSS::kSentinel) {
+            rev_seq.back() = complement(alphabet[c]);
+            node_index prev = graph_->kmer_to_node(rev_seq);
+            if (prev != DeBruijnGraph::npos)
+                callback(prev);
+        }
+    }
```

Comment:
better do inside -`else`. It's a relatively small chunk compared to the one under `if`